federation/service/node.go

```diff
@@ -0,0 +1,203 @@
+		sysUser      service.UserService
+		tokenEncoder auth.TokenEncoder
+	}
+
+	NodeService interface {
+		With(ctx context.Context) NodeService
+
+		Create(n *types.Node, sharedDomain string) (*types.Node, error)
+		Pair(nodeID uint64) error
+		HandshakeInit(sourceNodeID, nodeIDB uint64, nodeURI, token string) error
+		HandshakeConfirm(nodeID uint64) error
+		HandshakeComplete(nodeID uint64, token string) error
+	}
+)
+
+func Node(ctx context.Context) NodeService {
+	rand.Seed(time.Now().UnixNano())
+	return (&node{
+		store: DefaultNgStore,
+
+		actionlog:    DefaultActionlog,
+		tokenEncoder: auth.DefaultJwtHandler,
+	}).With(ctx)
+}
+
+func (svc node) With(ctx context.Context) NodeService {
+	return &node{
+		ctx: ctx,
+
+		actionlog: svc.actionlog,
+
+		store:        svc.store,
+		sysUser:      service.DefaultUser.With(ctx),
+		tokenEncoder: auth.DefaultJwtHandler,
+	}
+}
+
+// @todo move myDomain to configuration
+func (svc node) Create(n *types.Node, myDomain string) (*types.Node, error) {
+	n.ID = id.Next()
+	n.Status = types.NodeStatusPending
+
+	ott := util.GenerateOTT()
+	i := util.EncodeURI(ott, myDomain, n.ID)
+	if n.NodeURI == "" {
+		n.NodeURI = i
+	}
+
+	// @todo store an initial Node entry
+	tmpNodeStore = append(tmpNodeStore, n)
+
+	return n, nil
+}
+
+func (svc node) Pair(nodeID uint64) error {
+	// @todo store
+	n := tmpNodeStore.FindByID(nodeID)
+	if n == nil {
+		return ErrorNodeNotFound
+	}
+
+	// Handle fenedrated user
+	uHandle := "federation_" + strconv.FormatUint(n.ID, 10)
+	u, _ := svc.sysUser.FindByHandle(uHandle)
+
+	if u == nil {
+		var err error
+
+		// Create a system user to service this node
+		u, err = svc.sysUser.Create(&stypes.User{
+			Email:          strconv.FormatUint(n.ID, 10) + "@federation.corteza",
+			Handle:         uHandle,
+			Kind:           stypes.FederationUser,
+			EmailConfirmed: true,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Generate JWT token for the federated user
+	t := svc.tokenEncoder.Encode(u)
+	// @todo remove
+	spew.Dump(u, t)
+
+	// Ping node A to request the handshake
+	// @todo...
+
+	return nil
+}
+
+func (svc node) HandshakeInit(nodeIDA, nodeIDB uint64, nodeURI, token string) error {
+	// @todo store
+	n := tmpNodeStore.FindByID(nodeIDA)
+	if n == nil {
+		return ErrorNodeNotFound
+	}
+
+	if n.NodeURI != nodeURI {
+		return ErrorInvalidNodeURI
+	}
+
+	// @todo store...
+	n.Token = token
+	n.SharedID = nodeIDA
+	n.Status = types.NodeStatusPairRequest
+	spew.Dump(n)
+
+	// Notify the node administrator about the request
+	// @todo
+
+	return nil
+}
+
+func (svc node) HandshakeConfirm(nodeID uint64) error {
+	// @todo store
+	n := tmpNodeStore.FindByID(nodeID)
+	if n == nil {
+		return ErrorNodeNotFound
+	}
+
+	// Handle fenedrated user
+	uHandle := "federation_" + strconv.FormatUint(n.ID, 10)
+	u, _ := svc.sysUser.FindByHandle(uHandle)
+
+	if u == nil {
+		var err error
+
+		// Create a system user to service this node
+		u, err = svc.sysUser.Create(&stypes.User{
```

Comment:
Same comments as above + can we tuck fed. user creation logic into one single fn?